source/communication_interfaces/src/sockets/UDPSocket.cpp

```diff
@@ -0,0 +1,135 @@
+    }
+  }
+
+  if (this->timeout_duration_sec_->get_value() > 0.0) {
+    timeval timeout{};
+    auto secs = std::floor(this->timeout_duration_sec_->get_value());
+    timeout.tv_sec = static_cast<long>(secs);
+    timeout.tv_usec = static_cast<long>((this->timeout_duration_sec_->get_value() - secs) * 1e6);
+    if (setsockopt(this->server_fd_, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout)) != 0) {
+      throw exceptions::SocketConfigurationException("Setting socket timeout failed.");
+    }
+  }
+}
+
+bool UDPSocket::recvfrom(const sockaddr_in& address, ByteArray& buffer) {
+  std::vector<char> local_buffer(this->buffer_size_->get_value());
+  auto receive_length = ::recvfrom(
+      this->server_fd_, local_buffer.data(), this->buffer_size_->get_value(), 0, (sockaddr*) &(address),
+      &(this->addr_len_));
+  if (receive_length < 0) {
+    return false;
+  }
+  local_buffer.at(receive_length) = 0;
+  buffer.reset();
+  buffer.load(local_buffer.data(), receive_length);
+  return true;
+}
+
+bool UDPSocket::sendto(const sockaddr_in& address, const ByteArray& buffer) const {
+  std::vector<char> local_buffer;
+  buffer.copy_to(local_buffer);
+  int send_length = ::sendto(
+      this->server_fd_, local_buffer.data(), local_buffer.size(), 0, (sockaddr*) &(address), this->addr_len_);
+  return send_length == static_cast<int>(local_buffer.size());
+}
+
+void UDPSocket::validate_and_set_parameter(const std::shared_ptr<ParameterInterface>& parameter) {
+  // TODO can we put that in ParameterMap?
+  if (this->parameters_.find(parameter->get_name()) == this->parameters_.end()) {
+    throw state_representation::exceptions::InvalidParameterException(
+        "Invalid parameter '" + parameter->get_name() + "' for class UDPSocket.");
+  }
+  this->assert_parameter_valid(parameter);
+  if (parameter->is_empty()) {
+    throw state_representation::exceptions::InvalidParameterException(
+        "Parameter '" + parameter->get_name() + "' cannot be empty.");
+  }
+  if (parameter->get_name() == "timeout_duration_sec" && parameter->get_parameter_value<double>() < 0.0) {
+    throw state_representation::exceptions::InvalidParameterException(
+        "Parameter 'timeout_duration_sec' cannot be negative.");
+  }
+  // TODO can we put that in ParameterMap?
+  switch (parameter->get_parameter_type()) {
+    case ParameterType::STRING:
+      this->parameters_.at(parameter->get_name())->set_parameter_value(parameter->get_parameter_value<std::string>());
+      break;
+    case ParameterType::INT:
+      this->parameters_.at(parameter->get_name())->set_parameter_value(parameter->get_parameter_value<int>());
+      break;
+    case ParameterType::DOUBLE:
+      this->parameters_.at(parameter->get_name())->set_parameter_value(parameter->get_parameter_value<double>());
+      break;
+    default:
+      break;
+  }
+}
+
+void UDPSocket::close() {
+  if (this->server_fd_ >= 0) {
+    ::close(this->server_fd_);
```

Comment:
comment, nonblocking: this is an unfortunate syntax to resolve the base namespace, mainly just because it looks weird and confusing. I can't think of an alternative though.